source/endpoints_util.c

```diff
@@ -0,0 +1,548 @@
+        if (to_append.ptr[idx] == '"' && !(idx > 0 && to_append.ptr[idx - 1] == '\\')) {
+            ++*quote_count;
+        }
+    }
+    return aws_byte_buf_append_dynamic(buf, &to_append);
+}
+
+static struct aws_byte_cursor escaped_closing_curly = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("}}");
+static struct aws_byte_cursor escaped_opening_curly = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("{{");
+
+/*
+ * Small helper to deal with escapes correctly in strings that occur before
+ * template opening curly. General flow for resolving is to look for opening and
+ * then closing curly. This function correctly appends any escaped closing
+ * curlies and errors out if closing is not escaped (i.e. its unmatched).
+ */
+int s_append_template_prefix_to_buffer(
+    struct aws_byte_buf *out_buf,
+    struct aws_byte_cursor prefix,
+    size_t *quote_count) {
+    struct aws_byte_cursor split = {0};
+    struct aws_byte_cursor rest = {0};
+
+    while (s_split_on_first_delim(prefix, '}', &split, &rest)) {
+        if (s_buf_append_and_update_quote_count(out_buf, split, quote_count)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_GENERAL, "Failed to append to resolved template buffer.");
+            goto on_error;
+        }
+
+        if (*quote_count % 2 == 0) {
+            if (aws_byte_buf_append_byte_dynamic(out_buf, '}')) {
+                AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_GENERAL, "Failed to append to resolved template buffer.");
+                goto on_error;
+            }
+            aws_byte_cursor_advance(&rest, 1);
+            prefix = rest;
+            continue;
+        }
+
+        if (aws_byte_cursor_starts_with(&rest, &escaped_closing_curly)) {
+            if (aws_byte_buf_append_byte_dynamic(out_buf, '}')) {
+                AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_GENERAL, "Failed to append to resolved template buffer.");
+                goto on_error;
+            }
+            aws_byte_cursor_advance(&rest, 2);
+        } else {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_GENERAL, "Unmatched or unescaped closing curly.");
+            goto on_error;
+        }
+
+        prefix = rest;
+    }
+
+    if (s_buf_append_and_update_quote_count(out_buf, split, quote_count)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_GENERAL, "Failed to append to resolved template buffer.");
+        goto on_error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+    return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_RESOLVE_FAILED);
+}
+
+int aws_byte_buf_init_from_resolved_templated_string(
+    struct aws_allocator *allocator,
+    struct aws_byte_buf *out_buf,
+    struct aws_byte_cursor string,
+    aws_endpoints_template_resolve_fn resolve_callback,
+    void *user_data,
+    bool is_json) {
+    AWS_PRECONDITION(allocator);
+
+    struct aws_string *resolved_template = NULL;
+
+    if (aws_byte_buf_init(out_buf, allocator, string.len)) {
+        AWS_ZERO_STRUCT(out_buf);
```

Comment:
super trivial: don't need this, the init will actually do it for you on failure.